libs/ssl-config/src/main/java/org/elasticsearch/common/ssl/SslDiagnostics.java

```diff
@@ -0,0 +1,373 @@
+                        message.append(" {trusted issuer}");
+                    }
+                }
+                message.append(") ");
+            }
+            final X509Certificate root = chain[chain.length - 1];
+            if (isSelfIssued(root)) {
+                message.append("which is ").append(describeSelfIssuedCertificate(root, contextName, trustedIssuers));
+            } else {
+                final String rootIssuer = root.getIssuerX500Principal().getName();
+                message.append("which is issued by [")
+                    .append(rootIssuer)
+                    .append("] (but that issuer certificate was not provided in the chain)")
+                    .append(describeIssuerTrust(contextName, trustedIssuers, root, rootIssuer));
+
+            }
+        }
+        return message.toString();
+    }
+
+    private static CharSequence describeIssuerTrust(String contextName, @Nullable Map<String, List<X509Certificate>> trustedIssuers,
+                                                    X509Certificate certificate, String issuerName) {
+        if (trustedIssuers == null) {
+            return "";
+        }
+        StringBuilder message = new StringBuilder();
+        final IssuerTrust trust = checkIssuerTrust(trustedIssuers, certificate);
+        if (trust.isVerified()) {
+            message.append("; the issuing certificate is trusted in this ssl context ([")
+                .append(contextName)
+                .append("]) with ")
+                .append(fingerprintDescription(trust.issuerCerts));
+        } else if (trust.foundCertificateForDn()) {
+            message.append("; this ssl context ([")
+                .append(contextName)
+                .append("]) trusts [")
+                .append(trust.issuerCerts.size())
+                .append("] ").append(trust.issuerCerts.size() == 1 ? "certificate" : "certificates")
+                .append(" with subject name [")
+                .append(issuerName)
+                .append("] and ")
+                .append(fingerprintDescription(trust.issuerCerts))
+                .append(" but the signatures do not match");
+        } else {
+            message.append("; this ssl context ([")
+                .append(contextName)
+                .append("]) is not configured to trust that issuer");
+        }
+        return message;
+    }
+
+    private static CharSequence describeSelfIssuedCertificate(X509Certificate certificate, String contextName,
+                                                              @Nullable Map<String, List<X509Certificate>> trustedIssuers) {
+        final StringBuilder message = new StringBuilder();
+        final CertificateTrust trust = isTrusted(trustedIssuers, certificate);
+        message.append("self-issued [").append(certificate.getIssuerX500Principal().getName()).append("] and is ")
+            .append(trust.isTrusted() ? "trusted" : "not trusted")
```

Comment:
I've changed it to

> the certificate is signed by (subject [CN=Test CA 1] fingerprint [2b7b0416391bdf86502505c23149022d2213dadc] {trusted issuer}) which is self-issued; the [CN=Test CA 1] certificate is trusted in this ssl context ([xpack.security.authc.realms.saml.saml1.ssl]) because we trust a certificate with fingerprint [1f8ac10f23c5b5bc1167bda84b833e5c057a77d2] for the same public key